src/module.c

```diff
@@ -647,27 +655,36 @@ static void freeRedisModuleAsyncRMCallPromise(RedisModuleAsyncRMCallPromise *pro
     zfree(promise);
 }
 
+static void moduleFreeArgv(robj **argv, int argc) {
+    for (int j = 0; j < argc; j++)
+        decrRefCount(argv[j]);
+    zfree(argv);
+}
+
 void moduleReleaseTempClient(client *c) {
-    if (moduleTempClientCount == moduleTempClientCap) {
-        moduleTempClientCap = moduleTempClientCap ? moduleTempClientCap*2 : 32;
-        moduleTempClients = zrealloc(moduleTempClients, sizeof(c)*moduleTempClientCap);
-    }
-    clearClientConnectionState(c);
-    listEmpty(c->reply);
-    c->reply_bytes = 0;
-    c->duration = 0;
-    resetClient(c);
-    c->bufpos = 0;
-    c->flags = CLIENT_MODULE;
-    c->user = NULL; /* Root user */
-    c->cmd = c->lastcmd = c->realcmd = NULL;
-    if (c->bstate.async_rm_call_handle) {
-        RedisModuleAsyncRMCallPromise *promise = c->bstate.async_rm_call_handle;
-        promise->c = NULL; /* Remove the client from the promise so it will no longer be possible to abort it. */
-        freeRedisModuleAsyncRMCallPromise(promise);
-        c->bstate.async_rm_call_handle = NULL;
+    if (c->flags & CLIENT_MODULE_PERSISTENT) {
+        /* if a persistent client, only reset flags and free argv
+         * don't reset client in general or return to pool */
+        c->flags &= ~(CLIENT_DENY_BLOCKING|CLIENT_ALLOW_OOM|CLIENT_READONLY|CLIENT_ASKING);
```

Comment:
these the flags set in RM_Call().  I figured flags set elsewhere are "runtime" flags (as will be set due to call() execution)

Reply:
ok, it may not help much, we can add a comment in RM_Call saying that each flag that's set there should be also added to this mask.

the other alternative is to add another field to the client struct, which is a mask of flags to be unset, and then for each flag we add to the `c->flags` we'll also add it to `c->flags_to_clear`.

WDYT?

Reply:
going with the reset_flag approach and going to add a setClientRMCallFlags(c, uint64_t) function that all flags should be set through (to set it in both)